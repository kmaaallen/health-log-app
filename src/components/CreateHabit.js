import React, { useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTheme, Button, Dialog, TextInput, Portal, Surface, HelperText } from 'react-native-paper';
//REDUX
import { connect } from 'react-redux';
import { createHabit } from '../redux/actions';
//FORM
import { useForm, Controller } from "react-hook-form";
//COMPONENTS
import { Picker } from "@react-native-picker/picker";

const styles = theme => StyleSheet.create({
    surface: {
        padding: theme.spacing.medium,
        margin: theme.spacing.medium,
    },
    pickerView: {
        borderWidth: 1,
        borderColor: '#ccc',
        borderRadius: 4,
        height: 100
    }
});

//TODO CHECK HOW IT LOOKS ON ANDROID
//TODO STYLE ON ALL DEVICES
//TODO ALLOW SELECTION OF MULTIPLE CATEGORIES
//TODO ALLOW UPDATE OF CATEGORIES
//LOG PAGE RE-ORDER AND FILTER HABITS
//TODO SORT OUT MOBILE MODAL OVERLAP ISSUE

export const CreateHabit = (props) => {
    const [showDialog, setShowDialog] = useState(false);

    const { control, reset, handleSubmit, watch, formState: { errors }, register } = useForm({
        mode: 'onSubmit',
    });

    const watchCategoryPicker = watch("category", false);

    const createHabit = (formData) => {
        var category;
        formData.newCategory ? category = formData.newCategory : category = formData.category;
        props.createHabit(formData.title, formData.limit, category);
        setShowDialog(false);
        reset();
    }

    return (
        <Surface elevation={4} style={styles(props.theme).surface}>
            <Button onPress={() => setShowDialog(true)}>Create new habit</Button>
            <Portal>
                <Dialog visible={showDialog} onDismiss={() => setShowDialog(false)}>
                    <Dialog.Title>Create a habit</Dialog.Title>
                    <Dialog.Content>
                        <Controller
                            control={control}
                            render={({ field: { onChange, value } }) => (
                                <TextInput
                                    label="Title"
                                    onChangeText={value => onChange(value)}
                                    value={value || ''} // to remove uncontrolled to controlled error
                                    testID="title-input"
                                />
                            )}
                            name="title"
                            rules={{ required: true }}
                        />
                        {errors.title && <HelperText type="error" >Title is required</HelperText>}
                        <Controller
                            control={control}
                            render={({ field: { onChange, value } }) => (
                                <TextInput
                                    label="Daily limit"
                                    keyboardType="numeric"
                                    onChangeText={value => onChange(value)}
                                    value={value || ''}
                                    testID="limit-input"
                                />
                            )}
                            name="limit"
                            rules={{ required: true, min: 1 }}
                        />
                        {errors.limit && <HelperText type="error">{errors.limit.type == 'min' ? 'Limit must be greater than zero' : 'Limit is required'}</HelperText>}

                        <Controller
                            control={control}
                            render={({ field: { onChange, value } }) => (
                                <View style={styles(props.theme).pickerView}>
                                    <Picker
                                        selectedValue={value}
                                        onValueChange={value => onChange(value)}
                                        itemStyle={{ fontSize: 16, height: 100 }}
                                        testID="category-input"
                                    >
                                        <Picker.Item label="Please select a category" value="" key="empty" testID="select-category" />
                                        <Picker.Item label="New category" value="New category" key="New category" testID="new-category" />
                                        {props.categories.map((item) => (<Picker.Item key={item} label={item} value={item} />))}
                                    </Picker>
                                </View>
                            )}
                            name="category"
                            rules={{ required: false }}
                        />
                        {watchCategoryPicker == 'New category' ? <Controller
                            control={control}
                            render={({ field: { onChange, value } }) => (
                                <TextInput
                                    label="New Category"
                                    onChangeText={value => onChange(value)}
                                    value={value || ''}
                                    testID="new-category"
                                    placeholder="Create a new category"
                                />
                            )}
                            name="newCategory"
                            rules={{ required: false }}
                        /> : null}
                    </Dialog.Content>
                    <Dialog.Actions>
                        <Button onPress={() => setShowDialog(false)}>Cancel</Button>
                        <Button onPress={handleSubmit(createHabit)}>Ok</Button>
                    </Dialog.Actions>
                </Dialog>
            </Portal>
        </Surface >

    );
};

const mapStateToProps = (state) => ({
    categories: [...new Set((Object.values(state.habits.habits)).map((item) => item.category))].filter(item => item != "" && item != undefined).sort(),
})

const mapDispatchToProps = (dispatch) => {
    return {
        createHabit: (title, limit, category) => dispatch(createHabit({ updated: (new Date()).valueOf(), title: title, limit: limit, category: category }))
    }
}

export default withTheme(connect(mapStateToProps, mapDispatchToProps)(CreateHabit));