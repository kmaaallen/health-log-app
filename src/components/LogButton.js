import React, { Component } from 'react';
import { Card, Button, Title, Paragraph, Dialog, TextInput } from 'react-native-paper';
import { StyleSheet } from 'react-native';
//REDUX
import { connect } from 'react-redux';
import { incrementCount, resetCount, setLimit } from '../redux/actions';
import { hasReachedDailyLimitSelector } from '../redux/selectors';

const styles = StyleSheet.create({
    container: {
        height: 150,
    }
});

//TODO: NUMBER VALIDATION

export class LogButton extends Component {
    constructor(props) {
        super(props);
        this.state = { visible: false, tempLimit: 0 };
        this.showDialog = this.showDialog.bind(this);
        this.hideDialog = this.hideDialog.bind(this);
        this.setLimit = this.setLimit.bind(this);
    }

    showDialog = () => this.setState({ visible: true });

    hideDialog = () => this.setState({ visible: false });

    setTempLimit = (limit) => { this.setState({ tempLimit: limit }) };

    setLimit = () => {
        this.props.setLimit(this.state.tempLimit);
        this.hideDialog();
    }

    componentDidMount() {
        var midnight = new Date();
        midnight.setUTCHours(0, 0, 0, 0);
        if (new Date(this.props.updated) < midnight) {
            this.props.resetCount();
        }
    }

    render() {
        return (
            <Card style={styles.container} id={this.props.id}>
                <Card.Content>
                    <Title>My Habit</Title>
                    <Paragraph>{this.props.habit.count} / {this.props.habit.limit}</Paragraph>
                    <Paragraph>Last logged: {this.props.updatedDisplay}</Paragraph>
                </Card.Content>
                <Card.Actions>
                    <Button mode='contained' disabled={this.props.hasReachedLimit} onPress={this.incrementCount}>+</Button>
                    <Button mode='contained' onPress={this.showDialog}>Set Limit</Button>
                </Card.Actions>
                <Dialog visible={this.state.visible} onDismiss={this.hideDialog}>
                    <Dialog.Title>Choose a daily limit</Dialog.Title>
                    <Dialog.Content>
                        <TextInput
                            label="Daily limit"
                            onChangeText={(limit) => this.setTempLimit(limit)}
                        />
                    </Dialog.Content>
                    <Dialog.Actions>
                        <Button onPress={this.hideDialog}>Cancel</Button>
                        <Button onPress={this.setLimit}>Ok</Button>
                    </Dialog.Actions>
                </Dialog>
            </Card >
        );
    }
};

function mapStateToProps(state, ownProps) {
    const habit = state.count.habits[ownProps.id];
    const lastIncrementLog = habit.log.filter((log) => { return log.info.type == 'increment' }).slice(-1);
    return {
        habit: habit,
        lastUpdated: habit ? lastIncrementLog.updated : null,
        updatedDisplay: lastIncrementLog ? (new Date(lastIncrementLog.updated)).toLocaleString() : 'Never',
        hasReachedLimit: habit ? hasReachedDailyLimitSelector(state.count, ownProps.id) : {},
    }
}

const mapDispatchToProps = (dispatch, ownProps) => {
    return {
        incrementCount: () => dispatch(incrementCount((new Date()).valueOf()), ownProps.id),
        resetCount: () => dispatch(resetCount((new Date()).valueOf(), ownProps.id)),
        setLimit: (limit) => dispatch(setLimit((new Date()).valueOf(), limit, ownProps.id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(LogButton);