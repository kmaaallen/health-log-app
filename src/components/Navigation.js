import React from 'react';
//Navigation
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { Ionicons } from '@expo/vector-icons';
//Screens
import LogPage from '../screens/LogPage';
import NewHabit from '../screens/NewHabit';
import EditHabit from '../screens/EditHabit';
import Report from '../screens/Report';
//Theme
import { withTheme } from 'react-native-paper';

const Stack = createNativeStackNavigator();
const Tab = createBottomTabNavigator();

export const TabNavigation = () => {
    return (
        <Tab.Navigator
            screenOptions={({ route }) => ({
                tabBarIcon: ({ focused, color, size }) => {
                    let iconName;
                    if (route.name === 'Log') {
                        iconName = focused ? 'list-circle' : 'list-circle-outline';
                    } else if (route.name === 'New') {
                        iconName = focused ? 'add-circle' : 'add-circle-outline';
                    }
                    return <Ionicons name={iconName} size={size} color={color} />;
                },
                //tabBarActiveTintColor: props.theme.colors.primary, to fix
                tabBarActiveTintColor: 'blue',
                tabBarInactiveTintColor: 'gray',
            })}
        >
            <Tab.Screen name="Log" component={LogPage} />
            <Tab.Screen name="New" component={NewHabit} />
        </Tab.Navigator>
    )
}

export const StackNavigation = (props) => {
    return (
        <Stack.Navigator>
            <Stack.Screen name="Health Log Tracker" component={TabNavigation} />
            <Stack.Screen name="Edit" component={EditHabit} />
            <Stack.Screen name="Report" component={Report} />
        </Stack.Navigator>
    )
}


export default withTheme(StackNavigation);

