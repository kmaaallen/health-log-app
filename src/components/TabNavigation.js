import React from 'react';
//Navigation
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { Ionicons } from '@expo/vector-icons';
//Screens
import LogPage from '../screens/LogPage';
import NewHabit from '../screens/NewHabit';
//Theme
import { withTheme } from 'react-native-paper';

const Tab = createBottomTabNavigator();

export const TabNavigation = (props) => {
    return (
        <Tab.Navigator
            screenOptions={({ route }) => ({
                tabBarIcon: ({ focused, color, size }) => {
                    let iconName;
                    if (route.name === 'Log') {
                        iconName = focused ? 'list-circle' : 'list-circle-outline';
                    } else if (route.name === 'New') {
                        iconName = focused ? 'add-circle' : 'add-circle-outline';
                    }
                    return <Ionicons name={iconName} size={size} color={color} />;
                },
                tabBarActiveTintColor: props.theme.colors.primary,
                tabBarInactiveTintColor: 'gray',
            })}
        >
            <Tab.Screen name="Log" component={LogPage} />
            <Tab.Screen name="New" component={NewHabit} />
        </Tab.Navigator>
    )
}


export default withTheme(TabNavigation);

