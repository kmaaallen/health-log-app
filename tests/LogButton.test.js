import React from 'react';
import renderer from 'react-test-renderer';

import { LogButton } from '../src/components/LogButton';


describe('<LogButton />', () => {
    it('has 1 child', () => {
        const tree = renderer.create(<LogButton />).toJSON();
        expect(tree.children.length).toBe(1);
    });
    it('renders correctly', () => {
        const tree = renderer.create(<LogButton />).toJSON();
        expect(tree).toMatchSnapshot();
    })
});
