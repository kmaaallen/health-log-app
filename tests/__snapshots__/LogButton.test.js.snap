// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogButton /> renders correctly 1`] = `
<View
  id={1}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderColor": "rgba(0, 0, 0, 0.12)",
      "borderRadius": 4,
      "elevation": 3,
      "margin": 20,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 3,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "paddingHorizontal": 16,
          },
          Object {
            "paddingTop": 16,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "500",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 20,
                  "letterSpacing": 0.15,
                  "lineHeight": 30,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        My First Habit
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        1
         / 
        1
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        Last logged: 
        2/2/2021, 03:04:05 AM
      </Text>
    </View>
    <View
      index={1}
      siblings={
        Array [
          "Card.Content",
          "Card.Actions",
          "withTheme(Dialog)",
        ]
      }
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "padding": 8,
          },
          undefined,
        ]
      }
      total={3}
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 0,
            "margin": 2,
            "minWidth": "auto",
            "width": 80,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    Object {
                      "marginHorizontal": 8,
                    },
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.32)",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              +
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#6200ee",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "minWidth": "auto",
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    Object {
                      "marginHorizontal": 8,
                    },
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Set Limit
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
